### PR TITLE
Minor LLVM improvements

### PR DESCRIPTION
**Pull request by RossComputerGuy**
## Description of changes

Make a few improvements / changes, doesn't make any rebuilds.

- As brought up by @philiptaron on Matrix, `llvmPackages_git` uses `recurseIntoAttrs` after `callPackage`. Since this isn't necessary, remove it.
- Expose `mkPackage` in `llvmPackagesSets` so other packages / libraries could "import" an unpackaged LLVM version.
- Add `recurseIntoAttrs` into each LLVM version attr sets for Hydra / NixOS Search.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by philiptaron** [comment]
I'm on mobile and can't see the link to the packages that hydra and pfborg now see due to the new recurseIntoAttrs. I see the 100-500 label. Could you link those for me?

**Comment by RossComputerGuy** [comment]
@philiptaron https://gist.github.com/GrahamcOfBorg/eb215c81dc1fa40f965275182774dcb4

File: pkgs/development/compilers/llvm/default.nix [hunk 1]
```diff
@@ -7,6 +7,7 @@
   stdenv,
   gcc12Stdenv,
   pkgs,
+  recurseIntoAttrs,
```
Comment (philiptaron) [inline pkgs/development/compilers/llvm/default.nix:10]: You can also reference this as [lib.recurseIntoAttrs](https://noogle.dev/f/lib/recurseIntoAttrs).

**Review by philiptaron — APPROVED** [review]

**Comment by philiptaron** [comment]
I think you ought to only recurseIntoAttrs for 16+. Building old compilers isn't a good use of hydra and unlikely to be relevant for humans.

**Comment by RossComputerGuy** [comment]
> I think you ought to only recurseIntoAttrs for 16+.

I'm pretty sure many packages still rely on LLVM versions 12 - 15 in Nixpkgs. I think it would be better if we were to make that change once more packages move away to newer versions. Will bring this up with the rest of the LLVM maintainers though.

**Comment by philiptaron** [comment]
> I'm pretty sure many packages still rely on LLVM versions 12 - 15 in Nixpkgs. I think it would be better if we were to make that change once more packages move away to newer versions. Will bring this up with the rest of the LLVM maintainers though.

All of those will still build, since if Nix builds something it'll obviously build all of its dependencies. They're building today, right? But what this influences are the packages like lldb where no other package takes a dependency on them. After this PR, Hydra will now build them where it previously didn't. And they'll show up in search results where they previously didn't.

**Comment by RossComputerGuy** [comment]
> Hydra will now build them where it previously didn't

Hydra was building them but when #325175 was merged, the `recurseIntoAttrs` was dropped by accident. This PR restores the `recurseIntoAttrs`.

**Comment by philiptaron** [comment]
I see. Well, up to you as to whether you and the other LLVM maintainers if want to restore the status quo ante. You've got acceptance with comments. 📬

**Comment by RossComputerGuy** [comment]
CC @Ericson2314 for feedback

**Review by philiptaron — APPROVED** [review]

**Comment by name-snrl** [comment]
> * Expose `mkPackage` in `llvmPackagesSets` so other packages / libraries could "import" an unpackaged LLVM version.

Hey @RossComputerGuy, I'm a bit late here. But shouldn't `mkPackage` also be inherited in all-packages.nix? or how should it be invoked?

**Pull request merged**